### DISCUSSION TITLE
再议nslookup www.google.com 应答为www.google.com.lan的问题

### DISCUSSION THREAD

**Discussion by liumingan** [nikkinikki-org/OpenWrt-nikki#116] [Q&A]
如果禁止挟持dns，而是使用dnsmasq转发的话，就没有这个问题。
![image](https://github.com/user-attachments/assets/ced00106-43e5-4b35-89bc-36cd62519e90)
不知道中间有什么区别，建议大佬添加使用dnsmasq的选项，自主选择是防火墙挟持，还是使用dnsmasq
转发，（使用dnsmasq转发的话要禁止dnsmasq的缓存）

**Comment by morytyann** [comment]
> 禁止挟持dns，而是使用dnsmasq转发的话，就没有这个问题

dnsmasq看到是.lan的就不会转发到上游，所以没事，你不信你可以抓包看看。
你就把那个本地域名删掉不行吗，不要再问这个问题了好吗，我都想BLOCK你了

> 添加使用dnsmasq的选项

不加，说过了不会修改其他插件的配置

**Comment by morytyann** [comment]
#92 那边已经说的够明白了

1. Fake-IP不管你什么域名都会返回结果，哪怕是不存在的域名，你可以随便打一个试试
2. nslookup会先查询追加了你的`本地域名`的地址，OpenWrt默认是`.lan`，所以先查询了`www.google.com.lan`
3. Mihomo返回了`www.google.com.lan`的Fake-IP结果

你说DNSMASQ转发没事
1. DNSMASQ收到`.lan`结尾的，知道是本地域名，没有转发到Mihomo
2. 查询之后发现没有，返回NX-DOMAIN
3. nslookup收到NX-DOMAIN就会查询正常的域名
4. 正常的域名返回正常的结果

解决办法
- 不要再用那个破nslookup，用PowerShell的`Resolve-DnsName`
- 如果你非要用，查询的时候在域名最后追加一个`.`（英文句号），例如`www.google.com.`
- 删掉那个`DHCP/DNS -> 常规 -> 本地域名`，保存并应用后，客户端重新获取DHCP

我在用OpenClash的时候就遇到过这个问题，我发现他不影响我正常上网，就猜到了这是nslookup自己的行为，不管它就行了，为何要纠结这问题？

@ForestL18 @zutzo 你俩能看明白吧，不是我说不清楚吧？

**Reply by ForestL18** [reply]
@morytyann 啥情况？我没提过这个问题啊🤔

**Reply by morytyann** [reply]
没，就是想让你看看是我说的不清楚吗……

**Reply by ForestL18** [reply]
OK，没问题啊老铁。

**Reply by ForestL18** [reply]
DNSMasq的默认行为和Windows软件的默认行为本身都没问题，但是fakeip的加入就打破了这个平衡🤣

**Reply by ForestL18** [reply]
nslookup：不是哥们？设计者也没说过我是被设计查fakeip啊😂

**Reply by zutzo** [reply]
我用的 DNSMASQ 转发 因为我要塞 Adgh 
而且 直接 劫持的话 服务器那里会显示 Unknown () 我有强迫症(

**Reply by ForestL18** [reply]
> 没，就是想让你看看是我说的不清楚吗……

大佬，问个题外话，我看你用的也是pwsh7，我这边首次启动pwsh7都很慢，有啥解决方案吗？😥

**Reply by morytyann** [reply]
我的之前有一段时间启动很慢，后来不知道是更新了还是怎么回事，正常了

**Comment by liumingan** [comment]
Joseph Mory ***@***.***> 于2024年8月24日周六 22:10写道：

> 我的之前有一段时间启动很慢，后来不知道是更新了还是怎么回事，正常了
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/morytyann/OpenWrt-mihomo/discussions/116#discussioncomment-10437936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHST537AJ4FAEFMYVF4I4IDZTCH6XAVCNFSM6AAAAABNBRMEBKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANBTG44TGNQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>
好吧，不再提这个问题了，不过，并不是仅仅在fakeip，在redir-host模式下也是这样的。
这个问题不再提了
